mineinabyss-features/src/main/kotlin/com/mineinabyss/armorstandlock/ArmorStandLockingListener.kt

```diff
@@ -92,4 +99,114 @@ class ArmorStandLockingListener : Listener {
             isCancelled = true
         }
     }
+
+    @EventHandler
+    fun EntityMoveEvent.onEntityMove() {
+        /**
+         * This listener is used to prevent gravity and water from affecting armor stands.
+         * This is done by setting the entity's velocity to 0 and cancelling the event.
+         */
+        if (entityType != EntityType.ARMOR_STAND) return
+
+        val armorStand = entity.toGeary().get<LockArmorStand>() ?: return
+
+        if (!armorStand.lockState) return
+
+        // We only care about ArmorStands that have fishing hooks as passengers
+        if (entity.passengers.none { it.type == EntityType.FISHING_HOOK }) return
+
+        entity.velocity = Vector(0.0, 0.0, 0.0)
+        isCancelled = true
+    }
+
+    @EventHandler
+    fun PlayerFishEvent.onFishArmorstand() {
+        /**
+         * This listener is used to prevent players from fishing armor stands if they are not allowed to.
+         * This is done by simply cancelling the event. A player can fish an armor stand if they are allowed to via /mia lock
+         * or if they have the permission to bypass the lock.
+         */
+        if (state != PlayerFishEvent.State.CAUGHT_ENTITY) return // We only care about catching entities
+
+        if (caught == null) return // We only care about non-null caught entities
+        if (caught!!.type != EntityType.ARMOR_STAND) return // We only care about armor stands
+
+        val armorStandEntity = caught as ArmorStand
+        val armorStand = armorStandEntity.toGeary().get<LockArmorStand>() ?: return
+
+        if (!armorStand.lockState) return
+
+        // The player is allowed to fish if they are the owner of the armor stand or if they have the permission to bypass the lock
+
+        if (!armorStand.isAllowed(player.uniqueId) && !player.hasPermission("mineinabyss.lockarmorstand.bypass")) {
+            player.error("You do not have access to fishing this armor stand!")
+            hook.remove()
+            isCancelled = true
+        }
+    }
+
+    private fun onPistonExtendRetract(block: Block, direction: BlockFace, blocks: List<Block>): Boolean {
+        /**
+         * This function is used to prevent pistons from extending armor stands,
+         * to prevent pistons from pushing blocks into armor stands,
+         * and to prevent pistons from pushing armor stands using honey blocks below them.
+         */
+
+        val blocksThePistonChanges = mutableListOf<Block>()
```

Comment:
In general, we dislike making things mutable when they shouldn't be, the idiomatic way would be to chain some function calls. I recommend getting familiar with Kotlin's [collections](https://kotlinlang.org/docs/collections-overview.html) and their operations. You don't need to do it here, but it could be a good exercise.

Reply:
I don't really see a way here to transform this into a immutable list. How would you do that?

Reply:
The the collections api, you could do something like this
```kt
val blocksThePistionChanges = (blocks + block).map { it.location.add(direction.direction).block } +
    blocks.filter { it.type == Material.HONEY+BLOCK }
        .map { it.location.add(BlockFace.UP.direction).block }
```
There's a ton of functions and you might be able to do something cleaner, but it's mainly a good exercise to start thinking this way now so you can build up *really* long chains in the future.

Notice also, the performance of this is technically worse, it does extra list creations and whatnot. When we get a big enough list and performance is a concern, Kotlin has an identical sequence api that will do operations in-place, but it adds enough overhead that it shouldn't be used for lists this small.

Reply:
I didn't know that works. Thanks a lot again!